frontend/server/src/main/java/org/pytorch/serve/wlm/SequenceBatchAggregator.java

```diff
@@ -0,0 +1,308 @@
+            // Send the error message to all the jobs
+            for (Map.Entry<String, Job> j : jobs.entrySet()) {
+                String jobsId = j.getValue().getJobId();
+                Job job = jobs.get(jobsId);
+
+                if (job.isControlCmd()) {
+                    job.sendError(status, error);
+                } else {
+                    // Data message can be handled by other workers.
+                    // If batch has gone past its batch max delay timer?
+                    if (job.getGroupId() == null) {
+                        model.addFirst(job);
+                    } else {
+                        logger.error(
+                                "Failed to process requestId: {}, sequenceId: {}",
+                                job.getPayload().getRequestId(),
+                                job.getGroupId());
+                    }
+                }
+            }
+        }
+        jobs.clear();
+    }
+
+    private void pollJobGroup() throws InterruptedException {
+        cleanJobGroup();
+        LinkedHashSet<String> tmpJobGroups = new LinkedHashSet<>();
+        if (jobGroups.size() == 0) {
+            String jobGroupId =
+                    model.getPendingJobGroups().poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+            jobGroups.add(jobGroupId);
+            logger.debug("added jobGroup:{} in jobGroups", jobGroupId);
+        }
+        int quota = model.getBatchSize() - jobGroups.size();
+        if (quota > 0 && model.getPendingJobGroups().size() > 0) {
+            model.getPendingJobGroups().drainTo(tmpJobGroups, quota);
+            jobGroups.addAll(tmpJobGroups);
+        }
+        if (!tmpJobGroups.isEmpty()) {
+            logger.debug(
+                    "added jobGroup:{} in jobGroups",
+                    tmpJobGroups.stream().map(String::valueOf).collect(Collectors.joining(",")));
+        }
+    }
+
+    public static <T> CompletableFuture<Void> allOfFutures(List<CompletableFuture<T>> futures) {
+        CompletableFuture<Void> allDoneFuture;
+        CompletableFuture<?>[] completableFuturesArray =
+                futures.toArray(new CompletableFuture<?>[0]);
+        allDoneFuture = CompletableFuture.allOf(completableFuturesArray);
+
+        return allDoneFuture;
+    }
+
+    private void pollJobFromGroups() throws InterruptedException, ExecutionException {
+        while (jobs.isEmpty()) {
+            if (jobGroups.size() < model.getBatchSize()) {
+                pollJobGroup();
+            }
+
+            List<CompletableFuture<Void>> futures = new ArrayList<>(jobGroups.size());
+            for (String groupId : jobGroups) {
+                JobGroup jobGroup = model.getJobGroup(groupId);
+                futures.add(
+                        CompletableFuture.runAsync(
+                                () -> {
+                                    Job job = jobGroup.pollJob((long) model.getMaxBatchDelay());
+                                    if (job != null) {
```

Comment:
Can you change this part into pushing the jobs instead of polling?

Reply:
The current logic is as the following:
- all the jobs of a sequence is pushed to the jobQ of this sequence's  jobGroup.
- sequence batch aggregator parallel poll a job from the jobQ of each jobGroup with timeout=maxBatchDelay.

Regarding to your comments
- poll a job from jobQ vs push a job from jobQ to a batch: usually batch aggregator choose poll mode b/c it is able to control or even filter the requests. Here, I still choose poll mode so that later continuous batching can be easily integrated into sequence batching too. Push mode makes it complicate for continuous batching to still process the previous job, not take the new job.
- delay from poll timeout=maxBatchDelay: I agree it can reduce the latency triggered by the different traffic volume from the different sequences. Without the option timeout=maxBatchDelay, it can also trigger high latency when there is a slightly difference in traffic pace among the sequences. For example, job A from sequence A comes to its jobQ at t1, job B from sequence B comes to its jobQ at t2=t1+1ms. Without timeout option, the batch aggregator creates a batchSize=1. In this case, job B has to wait for next batch time (ie. job A complete). This is the reason I keep the option timeout for poll.